ts/sagas/workUnit/index.ts

```diff
@@ -0,0 +1,107 @@
+import { NavigationActions, NavigationNavigateAction } from "react-navigation";
+import { call, Effect, put, select, take } from "redux-saga/effects";
+import {
+  ActionCreator,
+  ActionType,
+  getType,
+  isActionOf,
+  TypeConstant
+} from "typesafe-actions";
+import { navigationHistoryPop } from "../../store/actions/navigationHistory";
+import { navigationHistorySizeSelector } from "../../store/middlewares/navigationHistory";
+import { navigationCurrentRouteSelector } from "../../store/reducers/navigation";
+
+/**
+ * The data model needed to run the workunit
+ */
+export type WorkUnit = {
+  // The navigation action that will be used if the current screen isn't the `startScreenName`
+  startScreenNavigation: NavigationNavigateAction;
+  // The expected first screen of the workflow
+  startScreenName: string;
+  // The action that will be taken when the workflow is completed
+  complete: ActionCreator<TypeConstant>;
+  // The action that will be taken when the workflow has been canceled
+  cancel: ActionCreator<TypeConstant>;
+  // The action that will be taken when `back` is pressed from the `startScreenName`
+  back: ActionCreator<TypeConstant>;
+};
+
+/**
+ * The result of the WorkUnit
+ */
+export enum ESagaResult {
+  Cancel = "Cancel",
+  Completed = "Completed",
+  Back = "Back"
+}
+
+/**
+ * Ensure that the `startScreen` is the current screen or navigate to `startScreen` using `navigateTo`
+ * @param navigateTo
+ * @param startScreen
+ */
+function* ensureScreen(
+  navigateTo: NavigationNavigateAction,
+  startScreen: string
+) {
+  const currentRoute: ReturnType<typeof navigationCurrentRouteSelector> = yield select(
+    navigationCurrentRouteSelector
+  );
+
+  if (currentRoute.isSome() && currentRoute.value !== startScreen) {
+    yield put(navigateTo);
+  }
+}
+
+/**
+ * Ensure that after the execution of the saga `g`,
+ * the navigation stack return to the screen from which the saga was invoked
+ * @param g
+ */
+export function* withResetNavigationStack<T>(
```

Comment:
Could it be an enhancement?
In this way we can call `withResetNavigationStack` by passing also the parameters taken by `g`

```typescript
export function* withResetNavigationStack<T>(
  g: (...args: any) => Generator<Effect, T>,
  args?: Array<any>
): Generator<Effect, T, any> {
  const currentNavigationStackSize: ReturnType<typeof navigationHistorySizeSelector> = yield select(
    navigationHistorySizeSelector
  );
  const res: T = yield call(() => (args ? g(...args) : g()));
  const newNavigationStackSize: ReturnType<typeof navigationHistorySizeSelector> = yield select(
    navigationHistorySizeSelector
  );
  const deltaNavigation = newNavigationStackSize - currentNavigationStackSize;
  if (deltaNavigation > 1) {
    yield put(navigationHistoryPop(deltaNavigation - 1));
  }
  yield put(NavigationActions.back());
  return res;
}
```

Reply:
My idea, in case of a saga with some parameters was to use:
```typescript
  const res: SagaCallReturnType<typeof executeWorkUnit> = yield call(
    withResetNavigationStack,
    () => bancomatWorkUnit("Hola, ¿que tal?")
  );
```

In order to preserve the type of the arguments of the `g` method. Using `args?: Array<any>` I could pass any types and any size of parameters which may not match the signature.
What you think?